src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -103,6 +105,96 @@ internal sealed partial class LocalRewriter
 
                 return _factory.Sequence(temps.ToImmutableAndFree(), effects.ToImmutableAndFree(), returnValue);
             }
+
+            void reverseAssignmentsToTargetsIfApplicable()
+            {
+                PooledHashSet<Symbol>? visitedSymbols = null;
+
+                Debug.Assert(right is not ({ Kind: BoundKind.TupleLiteral } or BoundConversion { Operand.Kind: BoundKind.TupleLiteral }));
+                if (// was the RHS a tuple literal?
+                    right is { Kind: BoundKind.ConvertedTupleLiteral } or BoundConversion { Operand.Kind: BoundKind.ConvertedTupleLiteral }
+
+                    // were any of the RHS expressions stored into temps?
+                    && effects.init.Any()
+
+                    && canReorderTargetAssignments(lhsTargets, ref visitedSymbols))
+                {
+                    // Consider a deconstruction assignment like the following
+                    // (a, b, c) = (x, y, z);
+
+                    // (x, y, z) are evaluated into temps, then the temps are stored to the targets:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // a = temp1;
+                    // b = temp2;
+                    // c = temp3;
+
+                    // In this code path, we have found a case where a, b, and c are non-side-effecting expressions which refer to unique/non-aliased variables.
+                    // In addition, at least one of the expressions x, y, or z actually needed to be evaluated into temps (versus being a constant value where the temp was elided at an earlier phase).
+                    // As an optimization, ensure that assignments from temps to targets happen in the reverse order of effects:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // c = temp3;
+                    // b = temp2;
+                    // a = temp1;
+
+                    // This makes it more likely that the stack optimizer pass will be able to eliminate the temps and replace them with stack push/pops.
+                    // Note also that no optimization can occur if the expression being converted is not a literal, e.g. `(a, b, c) = M()`. Order of access won't affect our ability to eliminate temps in that case.
+                    effects.assignments.ReverseContents();
+                }
+
+                visitedSymbols?.Free();
+            }
+
+            static bool canReorderTargetAssignments(ArrayBuilder<Binder.DeconstructionVariable> targets, ref PooledHashSet<Symbol>? visitedSymbols)
+            {
```

Comment:
nit: Thanks for writing up the summary in OP. Consider capturing it here as comment:
```
Here are the general requirements for performing the optimization:
- the RHS is a tuple literal (which means the temps produced for this assignment are for the tuple elements, which could turn into push-pops into the destination variables)
- at least one element in the RHS is actually stored to a temp. i.e. it is not a constant expression.
- all variables on the LHS are unique, by-value, and are locals or parameters.
```

Reply:
Have adjusted the comments to include these, inline with the checks which correspond to these requirements.